src/lua/uri.lua

```diff
@@ -218,6 +253,50 @@ local function values(...)
     return {...}
 end
 
+local char_to_hex = function(c)
+    return string.format("%%%02X", string.byte(c))
+end
+
+-- Encodes a string into its escaped hexadecimal representation.
+local function str_escape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: str_escape(string, [opts])")
+    end
+    opts = opts or RFC3986
+    opts.unreserved = opts.unreserved or RFC3986.unreserved
+    opts.plus = opts.plus or RFC3986.plus
+
+    local url = buf
+    local pattern = ("([%s])"):format(opts.unreserved)
+    url = url:gsub("\n", "\r\n")
+    url = url:gsub(pattern, char_to_hex)
+    if opts.plus == true then
+        url = url:gsub(" ", "+")
+    end
+    return url
+end
+
+local hex_to_char = function(x)
+    return string.char(tonumber(x, 16))
+end
+
+-- Decodes a escaped hexadecimal string into its binary representation.
+local function str_unescape(buf, opts)
+    if type(buf) ~= "string" then
+        error("Usage: str_unescape(string, [opts])")
+    end
+    opts = opts or RFC3986
+    opts.unreserved = opts.unreserved or RFC3986.unreserved
+    opts.plus = opts.plus or RFC3986.plus
```

Comment:
This code may change user provided options. It may lead to hard-to-debug problems and undesirable. See an example of such problem: https://github.com/tarantool/crud/issues/192.

If you're in doubt, just follow this pattern:

```lua
local opts = opts or {}
local foo = opts.foo or foo_default
local bar = opts.bar or bar_default
```

Reply:
> to make you happy.

I'm happy when the new code don't introduce well known problems. As you see, I linked a user report regarding the same problem in crud: it is not mine report.

No, we shouldn't describe incorrect behavior in the documentation, we should fix it. Input parameters must not be changed. Output parameters may be changed.

Reply:
The code looks so now:

```lua
local options = opts or RFC3986
options.unreserved = options.unreserved or RFC3986.unreserved
options.plus = options.plus or RFC3986.plus
```

The first line doesn't perform a table copy (`table.copy()` does).

Let's look at the following example of a caller side code:

```lua
local opts = {plus = true}
local escaped = uri.escape(data, opts)
-- opts is now {plus = true, unreserved = uri.RFC3986.unreserved}
```

`uri.unescape()` also may change a user provided table.

----

There are different ways to write it without changing input arguments:

```lua
local opts = opts or RFC3986
local unreserved_arg = opts.unreserved or RFC3986.unreserved
-- Caution: `opts.x or default_opts.x` always return `true` if
-- `default_opts.x` is `true`. IOW, doesn't differentiate `false`
-- and `nil`.
assert(RFC3986.plus == false)
local plus_arg = opts.plus or RFC3986.plus
```

```lua
-- Caution: `opts.x or default_opts.x` always return `true` if
-- `default_opts.x` is `true`. IOW, doesn't differentiate `false`
-- and `nil`.
assert(RFC3986.plus == false)
local opts = {
    unreserved = (opts and opts.unreserved) or RFC3986.unreserved,
    plus = (opts and opts.plus) or RFC3986.plus,
}
```

```lua
-- Caution: `opts.x or default_opts.x` always return `true` if
-- `default_opts.x` is `true`. IOW, doesn't differentiate `false`
-- and `nil`.
assert(RFC3986.plus == false)
local opts = table.copy(opts or {})
opts.unreserved = opts.unreserved or RFC3986.unreserved
opts.plus = opts.plus or RFC3986.plus
```

```lua
local opts = fun.chain(RFC3986, opts or {}):tomap()
```

```lua
local function resolve_opts(opts, default)
    local res = {}
    for k, v in pairs(default) do
        if opts[k] ~= nil then
            res[k] = opts[k]
        else
            res[k] = v
        end
    end
    return res
end

local opts = resolve_opts(opts, RFC3986)
```

Some of them are concise, some are wordy. Some always create an extra table or copy an existing one, others don't do that for some cases. I'll accept any variant, but it shouldn't change input arguments.